### ISSUE TITLE
jinja template fails to render base64 unicode strings.

### ISSUE THREAD

**Issue by rene00** [saltstack/salt#16218]
Using [hashutil.base64_decodestring](http://docs.saltstack.com/en/latest/ref/modules/all/salt.modules.hashutil.html) to decode base64 unicode strings in a jinja template fails. I've tested with 2014.1.5, 2014.1.7 & 2014.7.0rc2.

states/base64_jinja_test/init.sls

```
base64_jinja_test:
  file.managed:
    - name: '/tmp/base64_jinja_test'
    - source: salt://base64_jinja_test/files/base64_jinja_test.jinja
    - template: jinja
```

base64_jinja_test/files/base64_jinja_test.jinja

```
# foo
{{ salt['hashutil.base64_decodestring']('Zm9v') }}
# foˆø
{{ salt['hashutil.base64_decodestring']('Zm/LhsO4') }}
```

`state.sls base64_jinja_test` fails with

```
[INFO    ] Running state [/tmp/base64_jinja_test] at time 01:56:38.480123
[INFO    ] Executing state file.managed for /tmp/base64_jinja_test
[DEBUG   ] Jinja search path: ['/srv/salt/states']
[DEBUG   ] Traceback (most recent call last):
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/states/file.py", line 1274, in managed
    **kwargs
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/modules/file.py", line 2099, in get_managed
    **kwargs)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 85, in render_tmpl
    output = render_str(tmplstr, context, tmplpath)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 303, in render_jinja_tmpl
    trace=tracestr)
SaltRenderError: Jinja error: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)
Traceback (most recent call last):
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
    output = jinja_env.from_string(tmplstr).render(**unicode_context)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 4, in top-level template code
UnicodeDecodeError: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)

; line 4

---
# foo
{{ salt['hashutil.base64_decodestring']('Zm9v') }}
# foˆø
{{ salt['hashutil.base64_decodestring']('Zm/LhsO4') }}    <======================

Traceback (most recent call last):
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
    output = jinja_env.from_string(tmplstr).render(**unicode_context)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
[...]

---

[ERROR   ] Unable to manage file: Jinja error: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)
Traceback (most recent call last):
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
    output = jinja_env.from_string(tmplstr).render(**unicode_context)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 4, in top-level template code
UnicodeDecodeError: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)

; line 4

---
# foo
{{ salt['hashutil.base64_decodestring']('Zm9v') }}
# foˆø
{{ salt['hashutil.base64_decodestring']('Zm/LhsO4') }}    <======================

Traceback (most recent call last):
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
    output = jinja_env.from_string(tmplstr).render(**unicode_context)
  File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
[...]

---
[INFO    ] Completed state [/tmp/base64_jinja_test] at time 01:56:38.485144
[DEBUG   ] Loaded no_out as virtual quiet
[DEBUG   ] Loaded json_out as virtual json
[DEBUG   ] Loaded yaml_out as virtual yaml
[DEBUG   ] Loaded pprint_out as virtual pprint
local:
----------
          ID: base64_jinja_test
    Function: file.managed
        Name: /tmp/base64_jinja_test
      Result: False
     Comment: Unable to manage file: Jinja error: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)
              Traceback (most recent call last):
                File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
                  output = jinja_env.from_string(tmplstr).render(**unicode_context)
                File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
                  return self.environment.handle_exception(exc_info, True)
                File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 742, in handle_exception
                  reraise(exc_type, exc_value, tb)
                File "<template>", line 4, in top-level template code
              UnicodeDecodeError: 'ascii' codec can't decode byte 0xcb in position 2: ordinal not in range(128)

              ; line 4

              ---
              # foo
              {{ salt['hashutil.base64_decodestring']('Zm9v') }}
              # foˆø
              {{ salt['hashutil.base64_decodestring']('Zm/LhsO4') }}    <======================

              Traceback (most recent call last):
                File "/srv/virtualenv/salt/lib/python2.6/site-packages/salt/utils/templates.py", line 263, in render_jinja_tmpl
                  output = jinja_env.from_string(tmplstr).render(**unicode_context)
                File "/srv/virtualenv/salt/lib/python2.6/site-packages/jinja2/environment.py", line 969, in render
              [...]
              ---
     Changes:

Summary
------------
Succeeded: 0
Failed:    1
------------
Total:     1
```

```
# /srv/virtualenv/salt/bin/salt --versions           Salt: 2014.1.11
         Python: 2.6.6 (r266:84292, Jan 22 2014, 09:42:36)
         Jinja2: 2.7.3
       M2Crypto: 0.22
 msgpack-python: 0.4.2
   msgpack-pure: Not Installed
       pycrypto: 2.6.1
         PyYAML: 3.11
          PyZMQ: 14.3.1
            ZMQ: 4.0.4
```

**Comment by JaseFace** [comment]
I'm seeing this while running 2014.7.1 with a fix for a similar issue from https://github.com/saltstack/salt/issues/19173, but that turned out to not resolve it.  The data is stored in the pillar as base64 so it displays properly, and the output is binary data.

```
local:
    Data failed to compile:
----------
    Rendering SLS 'base:salt.master' failed: Jinja error: 'ascii' codec can't decode byte 0xbc in position 43: ordinal not in range(128)
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/salt/utils/templates.py", line 280, in render_jinja_tmpl
    output = template.render(**decoded_context)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 969, in render
    return self.environment.handle_exception(exc_info, True)
  File "/usr/lib/python2.7/dist-packages/jinja2/environment.py", line 742, in handle_exception
    reraise(exc_type, exc_value, tb)
  File "<template>", line 71, in top-level template code
  File "/usr/lib/python2.7/dist-packages/jinja2/filters.py", line 442, in do_indent
    rv = (u'\n' + indention).join(s.splitlines())
UnicodeDecodeError: 'ascii' codec can't decode byte 0xbc in position 43: ordinal not in range(128)

; line 71

---
[...]
    - mode: 600
    - contents_newline: False
    - backup: minion
    - replace: False
    - contents: |
        {{ salt['hashutil.base64_decodestring'](salt['pillar.get']('secrets:gpgkeys:' + gpgfile)) | indent(8) }}    <======================
  {% endfor %}
[...]
---

**Comment by JaseFace** [comment]
Sorry to bump this one, but I wasn't certain if my trace was relevant to this issue, or if a new one would be appropriate.  If someone could point me in the right direction I might be able to assist in a fix here also.

**Comment by donatello** [comment]
The OP's issue (that I am also affected by) may be because of a limitation in Jinja - from their [docs](http://jinja.pocoo.org/docs/dev/api/#unicode):

> It is not possible to use Jinja2 to process non-Unicode data. The
> reason for this is that Jinja2 uses Unicode already on the
> language level. For example Jinja2 treats the non-breaking space
> as valid whitespace inside expressions which requires knowledge
> of the encoding or operating on an Unicode string.

I am trying to binary key files from pillar by storing them in base64 and decoding them in templates using something similar to:

```
{{ salt['hashutil.base64_decodestring'](salt['pillar.get']('spiped:key', '')) }}
```

Any possible workarounds?

**Comment by JaseFace** [comment]
@donatello You can store binary data directly into the pillar without encoding/decoding it and deploying it works great, but last I tested it, it would cause salt to crash when you it tried to render those pillar contents.

If it properly displayed binary pillar data, or likely better just output 'Binary data' instead of the payload itself, we could probably run with that solution and skip any encoding.

**Comment by terminalmage** [comment]
This should be resolved with 2018.3.0.
